### ISSUE TITLE
[feature] Do not store flags/env vars in lockfile

### ISSUE THREAD

**Issue by EugenMayer** [conan-io/conan#6338]
We had to find out, that environment variables like `CONAN_RUN_TESTS` or even custom flags like `--with-tests` are saved in the lockfile.

This is very unusual for such lockfiles over the whole sector ( npm/composer/bundle/berkshelf..) - i have never seen this being part of the lockfile.

The problem with that is, that we now need 2 lockfiles, one for the testrun and one for the production build. Having 2 lockfiles means we could end up having different dependencies in the tests then in the production build case, while this would be one of the core promises we want lockfiles to fullfil - "ship what you have build and tested".

I think there is no real reason for saving such ENV files / flags in the lockfile or at least, this should be an opt in feature ( or if really there is no other way .. and opt out feature ).

Thanks

**Comment by memsharded** [comment]
Hi @EugenMayer 

The fact is that most of the other build systems do not handle the same variability as Conan does. The dependency graph is different for each configuration, for every OS platform, compiler, settings, options... It is obvious that the graph resolves to different artifacts for libraries in Windows than in Linux, because binary compatibility, for example.

It is not only perfectly valid, but also relatively common to have something like this:

```python
def build_requirements(self):
      if with_tests:  # this could come from an environment variable or the --with-tests argument
             self.build_requires("gtest/1.1@conan/stable")
```
This dependency will be pulled and existing in the graph only if the tests are being built, thus the graph is different and the lockfile for it needs to be different. There are indeed 2 lockfiles.

As a consequence, the effective profile, including env-vars, need to be stored in the lockfiles, for reproducibility. A different thing might be a feature to be able to redefine/overwrite env-vars, for some exceptional cases, and always as an acknowledge risk accepted by the user.

Please let me know if this makes sense. Thanks!

**Comment by memsharded** [comment]
Conan 2.0 lockfiles do not store any profile at all, closing this ticket as solved in 2.0, thanks!
